contrib/prometheus-watcher/main.go

```diff
@@ -0,0 +1,291 @@
+		return err
+	}
+	defer d.Close()
+	names, err := d.Readdirnames(-1)
+	if err != nil {
+		return err
+	}
+	for _, name := range names {
+		s := filepath.Join(dir, name)
+		if s != tmpdir {
+			err = os.RemoveAll(s)
+			if err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+func (w *volumeWatcher) writeRuleConfigMap(rulesDir string, index int, configMap string) error {
+	configMapParts := strings.Split(configMap, "/")
+	if len(configMapParts) != 2 {
+		return fmt.Errorf("Malformatted configmap key: %s. Format must be namespace/name.", configMap)
+	}
+	configMapNamespace := configMapParts[0]
+	configMapName := configMapParts[1]
+
+	cm, err := w.client.CoreV1().GetConfigMap(context.TODO(), configMapName, configMapNamespace)
+	if err != nil {
+		return err
+	}
+
+	dir := filepath.Join(rulesDir, fmt.Sprintf("rules-%d", index))
+	err = os.MkdirAll(dir, os.ModePerm)
+	if err != nil {
+		return err
+	}
+
+	for filename, content := range cm.Data {
+		err = w.writeConfigMapFile(filepath.Join(dir, filename), content)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (w *volumeWatcher) writeConfigMapFile(filename, content string) error {
+	f, err := os.Create(filename)
+	if err != nil {
+		return err
+	}
+	defer f.Close()
+
+	_, err = f.WriteString(content)
+	return err
+}
+
+func (w *volumeWatcher) ReloadPrometheus() error {
+	req, err := http.NewRequest("POST", w.cfg.reloadUrl, nil)
+	if err != nil {
+		return err
+	}
+	resp, err := http.DefaultClient.Do(req)
+	if err != nil {
+		return err
+	}
+	resp.Body.Close()
+	if resp.StatusCode != 200 {
+		return fmt.Errorf("Received response code %s, expected 200", resp.StatusCode)
+	}
+	return nil
+}
+
+func (w *volumeWatcher) Refresh() {
+	w.logger.Log("msg", "Updating rule files...")
+	err := w.UpdateRuleFiles()
+	if err != nil {
+		w.logger.Log("msg", "Updating rule files failed.", "err", err)
+	} else {
+		w.logger.Log("msg", "Rule files updated.")
+	}
+
+	w.logger.Log("msg", "Reloading Prometheus...")
+	err = w.ReloadPrometheus()
+	if err != nil {
+		w.logger.Log("msg", "Reloading Prometheus failed.", "err", err)
+	} else {
+		w.logger.Log("msg", "Prometheus successfully reloaded.")
+	}
+}
+
+func (w *volumeWatcher) Run() {
+	watcher, err := fsnotify.NewWatcher()
+	if err != nil {
+		w.logger.Log("msg", "Creating a new watcher failed.", "err", err)
+		os.Exit(1)
+	}
+	defer watcher.Close()
+
+	done := make(chan bool)
+	go func() {
+		for {
+			select {
+			case event := <-watcher.Events:
+				if event.Op&fsnotify.Create == fsnotify.Create {
+					if filepath.Base(event.Name) == "..data" {
+						w.logger.Log("msg", "ConfigMap modified.")
+						w.Refresh()
+					}
+				}
+			case err := <-watcher.Errors:
+				w.logger.Log("err", err)
+			}
+		}
+	}()
+
+	w.logger.Log("msg", "Starting...")
+	w.Refresh()
+	err = watcher.Add(w.cfg.configVolumeDir)
+	if err != nil {
+		w.logger.Log("msg", "Adding config volume to be watched failed.", "err", err)
+		os.Exit(1)
+	}
+
+	<-done
```

Comment:
The done channel is never closed anywhere so there doesn't seem to be actual termination handling.
In this case we can just move the for loop down out of the go routine and down here, right?